### PR TITLE
Add big-endian support to TFLite FlatBuffers

### PR DESCRIPTION
**Pull request by kun-lu20**
This PR aims to achieve a holistic/generic solution for TFLite FlatBuffers endianness issue on s390x(big-endian machines). It follows the below guidelines on BE machines:

1. Provide the endianness conversion feature for constant buffers in a TFLite model, both in C++ and Python code.
2. Convert the buffers from LE(little-endian) to BE(big-endian) format when loading a TFLite model from a file.
3. Convert the buffers from BE to LE format when writing a serialized string of TFLite model to a file.
4. Keep the buffers in BE format when the model/buffer is in memory.

This PR won't cause regression on LE machines. After applying this patch, LE format serialized TFLite model files (.tflite or .bin) could be used across platforms with different endianness format.

The following test case failures could be fixed by this code change on BE machines:
 
```bash
--- In lite module:
//tensorflow/lite:signature_runner_test
//tensorflow/lite/c:c_api_signature_runner_test
//tensorflow/lite/c:c_test
//tensorflow/lite/experimental/acceleration/mini_benchmark:validator_runner_impl_test
//tensorflow/lite/experimental/acceleration/mini_benchmark:validator_runner_test
//tensorflow/lite/experimental/acceleration/mini_benchmark:validator_test
//tensorflow/lite/experimental/acceleration/mini_benchmark/metrics:blazeface_mlkit_v1_validation_regression_test
//tensorflow/lite/experimental/acceleration/mini_benchmark/metrics:mobilenet_float_validation_test
//tensorflow/lite/experimental/acceleration/mini_benchmark/metrics:mobilenet_quant_validation_test
//tensorflow/lite/experimental/acceleration/mini_benchmark/model_modifier:custom_validation_embedder_test
//tensorflow/lite/python:util_test
//tensorflow/lite/testing:zip_test_expand_dims_mlir-quant
//tensorflow/lite/testing:zip_test_fully_connected_mlir-quant
//tensorflow/lite/testing:zip_test_mean_mlir-quant
//tensorflow/lite/testing:zip_test_mirror_pad_mlir-quant
//tensorflow/lite/testing:zip_test_pad_mlir-quant
//tensorflow/lite/testing:zip_test_reduce_max_mlir-quant
//tensorflow/lite/testing:zip_test_reduce_min_mlir-quant
//tensorflow/lite/testing:zip_test_reshape_mlir-quant
//tensorflow/lite/testing:zip_test_resize_bilinear_mlir-quant
//tensorflow/lite/testing:zip_test_resize_bilinear_mlir-quant_xnnpack
//tensorflow/lite/testing:zip_test_resize_nearest_neighbor_mlir-quant
//tensorflow/lite/testing:zip_test_resize_nearest_neighbor_mlir-quant_xnnpack
//tensorflow/lite/testing:zip_test_slice_mlir-quant
//tensorflow/lite/testing:zip_test_softmax_mlir-quant
//tensorflow/lite/testing:zip_test_softmax_mlir-quant_xnnpack
//tensorflow/lite/testing:zip_test_split_mlir-quant
//tensorflow/lite/testing:zip_test_split_mlir-quant_xnnpack
//tensorflow/lite/testing:zip_test_squeeze_mlir-quant
//tensorflow/lite/testing:zip_test_strided_slice_mlir-quant
//tensorflow/lite/testing:zip_test_transpose_conv_mlir-quant
//tensorflow/lite/testing:zip_test_transpose_conv_mlir-quant_xnnpack
//tensorflow/lite/testing:zip_test_transpose_mlir-quant
//tensorflow/lite/testing:zip_test_where_mlir-quant
//tensorflow/lite/tools/optimize/calibration:calibrator_test
//tensorflow/lite/tools/optimize/debugging/python:debugger_test
//tensorflow/lite/tools/optimize/python:modify_model_interface_lib_test
//tensorflow/lite/python:lite_v2_test

--- In compiler module:
//tensorflow/compiler/mlir/lite/quantization/lite:quantize_model_test
//tensorflow/compiler/mlir/lite/quantization/lite:quantize_weights_test
//tensorflow/compiler/mlir/lite/tests/flatbuffer2mlir:constants.mlir.test
//tensorflow/compiler/mlir/lite/tests/flatbuffer2mlir:importer_test_min_max.cc.mlir.test
//tensorflow/compiler/mlir/lite/tests/flatbuffer2mlir:importer_test_min_max.cc.test
//tensorflow/compiler/mlir/lite/tests/flatbuffer2mlir:math.mlir.test
//tensorflow/compiler/mlir/lite/tests/flatbuffer2mlir:quantization.mlir.test
//tensorflow/compiler/mlir/lite/tests/flatbuffer2mlir:reshape.mlir.test
//tensorflow/compiler/mlir/lite/tests/flatbuffer2mlir:simple.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:bucketize.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:custom_op_with_tflite_op.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:depthwise_conv2d.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:depthwise_conv2d_v2.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:disable_flex_enable_builtin.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:flex_op_with_tflite_op.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:math.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:quantization.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:reshape.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:signature_def_output_override.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:simple.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:simple_with_connected_control_nodes.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:simple_with_unconnected_control_nodes.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:tf_entry_function.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:tfl_while_op.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:while_op.mlir.test
```

This PR will also fix issue #45009 and PR #57065.

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

### CODE REVIEW

**Comment by kun-lu20** [comment]
Hi @terryheo,

Hope all is well.

Since this PR will fix the TFLite FlatBuffers endianness issue discussed in https://github.com/tensorflow/tensorflow/issues/45009, could you please review it when you have some time?

Thank you very much!

**Comment by kun-lu20** [comment]
Hi @MeghnaNatraj ,

Could you please review this PR when you have some time?

Thank you very much!

File: tensorflow/compiler/mlir/lite/BUILD [hunk 1]
```diff
@@ -890,6 +890,7 @@ tf_native_cc_binary(
     srcs = ["flatbuffer_to_string.cc"],
     deps = [
         "//tensorflow/lite/schema:schema_fbs_with_reflection",
+        ":tensorflow_lite",
```
Comment (terryheo) [inline tensorflow/compiler/mlir/lite/BUILD:893]: Why is this needed?

File: tensorflow/lite/core/model_builder.cc [hunk 2]
```diff
@@ -90,6 +98,114 @@ std::unique_ptr<FlatBufferModel> FlatBufferModel::VerifyAndBuildFromBuffer(
...
 }
 
+void FlatBufferModel::ByteSwapSerializedModel(std::string* serialized_model){
+  const uint8_t* buffer =
+      reinterpret_cast<const uint8_t*>(serialized_model->c_str());
+  const tflite::Model* input_model = tflite::GetModel(buffer);
+  ByteSwapTFLiteModel(input_model);
+}
+
+void FlatBufferModel::ByteSwapBuffer(int8_t tensor_type, size_t buffer_size,
```
Comment (terryheo) [inline tensorflow/lite/core/model_builder.cc:108]: Is the target always writable?
What happens if the target memory is mmaped ?

File: tensorflow/lite/tools/serialization/writer_lib.cc [hunk 3]
```diff
@@ -69,6 +70,11 @@ TfLiteStatus WriteImpl(const std::string& filename, void* data, size_t size) {
   FILE* fp = fopen(filename.c_str(), "wb");
   if (!fp) return kTfLiteError;
 
+  if (!FLATBUFFERS_LITTLEENDIAN){
```
Comment (terryheo) [inline tensorflow/lite/tools/serialization/writer_lib.cc:73]: Who does define this?

File: tensorflow/lite/tools/test_utils.py [hunk 4]
```diff
@@ -117,7 +117,7 @@ def build_mock_flatbuffer_model():
   schema_fb.TensorStart(builder)
   schema_fb.TensorAddName(builder, string1_offset)
   schema_fb.TensorAddShape(builder, shape1_offset)
-  schema_fb.TensorAddType(builder, 0)
+  schema_fb.TensorAddType(builder, 3) #UINT8
```
Comment (terryheo) [inline tensorflow/lite/tools/test_utils.py:120]: for what?

File: tensorflow/compiler/mlir/lite/BUILD (hunk 1 above)
Comment (kun-lu20) [reply tensorflow/compiler/mlir/lite/BUILD:893]: This line was added here so that `FlatBufferModel::ByteSwapSerializedModel()` could be invoked in `tensorflow/compiler/mlir/lite/flatbuffer_to_string.cc`. 

The conversion could fix a series of `//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer` test cases in `compiler` module on BE machines.

File: tensorflow/lite/tools/serialization/writer_lib.cc (hunk 3 above)
Comment (kun-lu20) [reply tensorflow/lite/tools/serialization/writer_lib.cc:73]: `FLATBUFFERS_LITTLEENDIAN` is defined in the `flatbuffers` header file here: https://github.com/google/flatbuffers/blob/6f895f54c25aa19f5d84ac6cf7fa8bc955a14e1d/include/flatbuffers/base.h#L119-L139

**Comment by kun-lu20** [comment]
Hi @terryheo ,

Thanks for your feedback! I've replied to your comments, please take a look when you have some time.

File: tensorflow/lite/core/model_builder.cc (hunk 2 above)
Comment (kun-lu20) [reply tensorflow/lite/core/model_builder.cc:108]: Actually `FlatBufferModel::ByteSwapFlatBufferModel()` will allocate a new `FlatBufferModel` instance for byte swapping and return its handler.

**Comment by kun-lu20** [comment]
Hi @terryheo ,

Could you please review this PR again? Thank you very much!

**Comment by kun-lu20** [comment]
Hi @terryheo ,

I've solved the conflicts from master branch. Could you please do the review again when you have some time? 

Thank you!

File: tensorflow/lite/core/model_builder.cc (hunk 2 above)
Comment (terryheo) [reply tensorflow/lite/core/model_builder.cc:108]: Then will it increase memory consumption?

Random thoughts, how about creating a separate tool to update LE buffers to BE buffers?

File: tensorflow/lite/tools/serialization/writer_lib.cc (hunk 3 above)
Comment (terryheo) [reply tensorflow/lite/tools/serialization/writer_lib.cc:73]: Isn't it better to use "#ifdef FLATBUFFERS_LITTLEENDIAN" ?

File: tensorflow/lite/tools/test_utils.py (hunk 4 above)
Comment (terryheo) [reply tensorflow/lite/tools/test_utils.py:120]: Got it. Then `schema_fb.TensorAddType(builder, TensorType_UINT8)` ?

File: tensorflow/lite/core/model_builder.cc [hunk 5]
```diff
@@ -90,6 +98,114 @@ std::unique_ptr<FlatBufferModel> FlatBufferModel::VerifyAndBuildFromBuffer(
                                       error_reporter);
 }
 
+void FlatBufferModel::ByteSwapSerializedModel(std::string* serialized_model){
```
Comment (terryheo) [inline tensorflow/lite/core/model_builder.cc:101]: I also want to add conditional building (`#ifdef FLATBUFFERS_LITTLEENDIAN`) for all swapping functions which are only needed for BE targets.

File: tensorflow/compiler/mlir/lite/BUILD (hunk 1 above)
Comment (kun-lu20) [reply tensorflow/compiler/mlir/lite/BUILD:893]: Yes, I'll improve the code to use `model_builder` only. Thanks!

File: tensorflow/lite/core/model_builder.cc (hunk 2 above)
Comment (kun-lu20) [reply tensorflow/lite/core/model_builder.cc:108]: Yes, it will increase memory consumption on BE platforms, but only during the model loading/saving phase, so I think it will not incur big impact to the performance.

Creating a separate tool will also work, but it will bring extra steps such as using the tool to convert the .tflite files on BE platforms before running TensorFlow code. I think embedding the byte swapping code in TensorFlow code base would be more convenient for the users, thus they can use the same steps and .tflite files across LE and BE platforms, without the need to mark which .tflite files are in which endianness format or do the conversion manually. The endianness issue becomes transparent to the users.

File: tensorflow/lite/tools/serialization/writer_lib.cc (hunk 3 above)
Comment (kun-lu20) [reply tensorflow/lite/tools/serialization/writer_lib.cc:73]: Yes, I'll adjust the code accordingly. Thanks!

File: tensorflow/lite/tools/test_utils.py (hunk 4 above)
Comment (kun-lu20) [reply tensorflow/lite/tools/test_utils.py:120]: Yes, using `TensorType_UINT8` is better. I'll update the code. Thanks!

File: tensorflow/lite/core/model_builder.cc [hunk 6]
```diff
@@ -90,6 +98,114 @@ std::unique_ptr<FlatBufferModel> FlatBufferModel::VerifyAndBuildFromBuffer(
...
+  }
+}
+
+void FlatBufferModel::ByteSwapTFLiteModel(const tflite::Model* tfl_model){
+  for (size_t subgraph_idx = 0; subgraph_idx < tfl_model->subgraphs()->size();
+       subgraph_idx++) {
+    const tflite::SubGraph* subgraph = tfl_model->subgraphs()->Get(subgraph_idx);
+    for (size_t ts_idx = 0; ts_idx < subgraph->tensors()->size(); ts_idx++) {
+      const tflite::Tensor* tensor = subgraph->tensors()->Get(ts_idx);
+      if(tensor->buffer()>0 && tensor->buffer()<tfl_model->buffers()->size()){
```
[the remark this answers is not in the document]
Comment (kun-lu20) [reply tensorflow/lite/core/model_builder.cc:151]: I see. Will adjust the code to do swap only once. Thanks!

File: tensorflow/lite/core/model_builder.cc (hunk 5 above)
Comment (kun-lu20) [reply tensorflow/lite/core/model_builder.cc:101]: Yes, that makes sense. I'll add the macro conditional statement for them. Thank you very much for all the valuable comments!

File: tensorflow/compiler/mlir/lite/BUILD (hunk 1 above)
Comment (kun-lu20) [reply tensorflow/compiler/mlir/lite/BUILD]: Done.

File: tensorflow/lite/tools/serialization/writer_lib.cc (hunk 3 above)
Comment (kun-lu20) [reply tensorflow/lite/tools/serialization/writer_lib.cc]: Done.

File: tensorflow/lite/core/model_builder.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (kun-lu20) [reply tensorflow/lite/core/model_builder.cc]: Both the C++ code and Python code were updated to ensure the swapping would be done only once.

File: tensorflow/lite/core/model_builder.cc (hunk 5 above)
Comment (kun-lu20) [reply tensorflow/lite/core/model_builder.cc:105]: Done.

**Comment by kun-lu20** [comment]
Hi @terryheo ,

Thanks again for your valuable comments! I've updated the PR as per your suggestions, please review it again when you have some time.

**Comment by gbaned** [comment]
Hi @terryheo Can you please review this PR ? Thank you!

**Comment by kun-lu20** [comment]
Hi @terryheo ,

Happy New Year!

I've solved the conflicts from the master branch. Could you please review this PR again? Thank you very much!

**Comment by kun-lu20** [comment]
Hi @terryheo ,

I've solved the conflicts from the master branch. Could you please take a look at this PR?

Thank you very much!

File: tensorflow/compiler/mlir/lite/flatbuffer_to_string.cc [hunk 7]
```diff
@@ -137,6 +140,9 @@ int main(int argc, char** argv) {
 
   std::string serialized_model;
   if (tflite::ReadAndVerify(argv[1], &serialized_model)) return 1;
+  #if FLATBUFFERS_LITTLEENDIAN == 0
```
Comment (terryheo) [inline tensorflow/compiler/mlir/lite/flatbuffer_to_string.cc:143]: We don't indent for "#if" "#endif". Plz update all.

File: tensorflow/lite/core/model_builder.cc [hunk 8]
```diff
@@ -90,6 +100,124 @@ std::unique_ptr<FlatBufferModel> FlatBufferModel::VerifyAndBuildFromBuffer(
...
+      for (int i = 0; i < buffer_size/8; i++)
+        bp[i] = flatbuffers::EndianSwap(bp[i]);
+      break;
+    }
+    default:
+      break;
+  }
+}
+
+void FlatBufferModel::ByteSwapTFLiteModel(const tflite::Model* tfl_model){
```
Comment (terryheo) [inline tensorflow/lite/core/model_builder.cc:149]: Could you run formatter? https://www.tensorflow.org/community/contribute/code_style#c_coding_style

Need a space before '{'

File: tensorflow/compiler/mlir/lite/flatbuffer_to_string.cc (hunk 7 above)
Comment (kun-lu20) [reply tensorflow/compiler/mlir/lite/flatbuffer_to_string.cc]: Done. Thanks!

File: tensorflow/lite/core/model_builder.cc (hunk 8 above)
Comment (kun-lu20) [reply tensorflow/lite/core/model_builder.cc]: I've run the formatter and updated the code. Thanks!

**Comment by kun-lu20** [comment]
Hi @terryheo ,

Thanks for your feedback! I've run the formatter and updated the related code. 

Please take a look when you have some time. Thanks again!

**Comment by kun-lu20** [comment]
Hi @terryheo ,

Hope all is well.

Could you please review this PR again? Thank you very much!

**Comment by kun-lu20** [comment]
Hi @terryheo ,

Could you please have a look at the code changes when you have some time?

Thank you!

File: tensorflow/lite/core/model_builder.cc [hunk 9]
```diff
@@ -94,6 +105,124 @@ std::unique_ptr<FlatBufferModel> FlatBufferModel::VerifyAndBuildFromBuffer(
...
+      auto bp = reinterpret_cast<uint16_t*>(buffer);
+      for (int i = 0; i < buffer_size / 2; i++)
+        bp[i] = flatbuffers::EndianSwap(bp[i]);
+      break;
+    }
+    // 32-bit types
+    case tflite::TensorType_FLOAT32:
+    case tflite::TensorType_INT32:
+    case tflite::TensorType_UINT32:
+    case tflite::TensorType_COMPLEX64: {
```
Comment (terryheo) [inline tensorflow/lite/core/model_builder.cc:133]: TensorType_COMPLEX64 is 64bit

File: tensorflow/lite/core/model_builder.cc [hunk 10]
```diff
@@ -94,6 +105,124 @@ std::unique_ptr<FlatBufferModel> FlatBufferModel::VerifyAndBuildFromBuffer(
...
+      auto bp = reinterpret_cast<uint32_t*>(buffer);
+      for (int i = 0; i < buffer_size / 4; i++)
+        bp[i] = flatbuffers::EndianSwap(bp[i]);
+      break;
+    }
+    // 64-bit types
+    case tflite::TensorType_INT64:
+    case tflite::TensorType_FLOAT64:
+    case tflite::TensorType_UINT64:
+    case tflite::TensorType_COMPLEX128: {
```
Comment (terryheo) [inline tensorflow/lite/core/model_builder.cc:143]: TensorType_COMPLEX128 is 128bit

**Review by terryheo — COMMENTED** [review]
Mostly LGTM

File: tensorflow/lite/core/model_builder.cc (hunk 9 above)
Comment (terryheo) [reply tensorflow/lite/core/model_builder.cc:133]: More preciously it consists with two 32bit values.

Comment (kun-lu20) [reply tensorflow/lite/core/model_builder.cc:133]: Thanks @terryheo !

If I understood correctly, a TensorType_COMPLEX64 data is 64-bit but it's composed of two 32-bit integer numbers, so it's handled along with other 32-bit data types.

Please correct me if I was wrong. Thanks again!

File: tensorflow/lite/core/model_builder.cc (hunk 10 above)
Comment (kun-lu20) [reply tensorflow/lite/core/model_builder.cc:143]: The reason is the same as the above TensorType_COMPLEX64.

File: tensorflow/lite/core/model_builder.cc (hunk 9 above)
Comment (terryheo) [reply tensorflow/lite/core/model_builder.cc:133]: Oh, I see. You're right.

**Review by terryheo — APPROVED** [review]
LGTM.

If there is  an internal test error with this PR, I'll ask you to fix it.

**Comment by kun-lu20** [comment]
Sure, will do. Thanks @terryheo !

File: tensorflow/lite/core/model_builder.h [hunk 11]
```diff
@@ -155,6 +155,33 @@ class FlatBufferModel {
...
+  static void ByteSwapBuffer(int8_t tensor_type, size_t buffer_size,
+                             uint8_t* buffer);
+
+  /// Byte swap the buffers field of a TFLite Model instance in place.
+  static void ByteSwapTFLiteModel(const tflite::Model* tfl_model);
+
+  /// Byte swap the buffers field of a TFLite ModelT instance in place.
+  static void ByteSwapTFLiteModelT(tflite::ModelT* tfl_modelt);
+
+  /// Convert the TFLite buffers field between LE and BE format in a
```
Comment (terryheo) [inline tensorflow/lite/core/model_builder.h:169]: This file has some unnecessary ending white spaces. Could you remove all?

Comment (kun-lu20) [reply tensorflow/lite/core/model_builder.h:169]: Sure

Comment (kun-lu20) [reply tensorflow/lite/core/model_builder.h]: Hi @terryheo ,

I've used `clang-format` to remove all unnecessary ending white spaces from this file. Also I've fixed another format issue in `tensorflow/lite/tools/serialization/BUILD` according to the `Code Check - Changed Files` results.

Could you please take a look?  Thank you very much!

**Review by terryheo — APPROVED** [review]
LGTM

**Comment by kun-lu20** [comment]
Thanks @terryheo

**Comment by kun-lu20** [comment]
Hi @terryheo ,

I've fixed the conflicts from the master branch. Please take a look when you have some time. 

BTW, I found that the latest result from `feedback/copybara` showed the internal checks failed on `2023-02-16`. Are there any code/BUILD files I need to change? 

Thanks!

**Review by terryheo — APPROVED** [review]

**Comment by kun-lu20** [comment]
Hi @terryheo ,

I saw all the checks have passed. Wondering if it's possible for this PR to be merged?

Thank you very much!

File: tensorflow/lite/tools/flatbuffer_utils.py [hunk 12]
```diff
@@ -274,6 +281,78 @@ def xxd_output_to_object(input_cc_file):
   return convert_bytearray_to_object(model_bytes)
 
 
+def byte_swap_buffer_content(buffer, chunksize, from_endiness, to_endiness):
+  """"Helper function for byte-swapping the buffers field.
```
Comment (terryheo) [inline tensorflow/lite/tools/flatbuffer_utils.py:285]: """Helper function for byte-swapping the buffers field."""

1. make it one line
2. there are 4 double quotes.

File: tensorflow/lite/tools/flatbuffer_utils.py [hunk 13]
```diff
@@ -274,6 +281,78 @@ def xxd_output_to_object(input_cc_file):
...
+    byteswap, from_endiness).to_bytes(
+      chunksize, to_endiness) for byteswap in to_swap])
+
+
+def byte_swap_tflite_model_obj(model, from_endiness, to_endiness):
+  """"Byte swaps the buffers field in a TFLite model.
+
+  Args:
+    model: TFLite model object of from_endiness format.
+
```
Comment (terryheo) [inline tensorflow/lite/tools/flatbuffer_utils.py:299]: you need to describe every arguments

File: tensorflow/lite/tools/flatbuffer_utils.py [hunk 14]
```diff
@@ -274,6 +281,78 @@ def xxd_output_to_object(input_cc_file):
...
+  """
+  to_swap = [buffer.data[i:i+chunksize] for i in range(
+    0, len(buffer.data), chunksize)]
+  buffer.data = b''.join([int.from_bytes(
+    byteswap, from_endiness).to_bytes(
+      chunksize, to_endiness) for byteswap in to_swap])
+
+
+def byte_swap_tflite_model_obj(model, from_endiness, to_endiness):
+  """"Byte swaps the buffers field in a TFLite model.
```
Comment (terryheo) [inline tensorflow/lite/tools/flatbuffer_utils.py:295]: plz use 3 double quotes instead of 4

File: tensorflow/lite/tools/flatbuffer_utils.py [hunk 15]
```diff
@@ -274,6 +281,78 @@ def xxd_output_to_object(input_cc_file):
...
+        elif tensor.type in types_of_64_bits:
+          byte_swap_buffer_content(model.buffers[tensor.buffer], 
+            8, from_endiness, to_endiness)
+        else:
+          continue
+        buffer_swapped.append(tensor.buffer)
+
+
+def byte_swap_tflite_buffer(tflite_model, from_endiness, to_endiness):
+  """"Generates a new model byte array after byte swapping its buffers field.
```
Comment (terryheo) [inline tensorflow/lite/tools/flatbuffer_utils.py:333]: use 3 double quotes

File: tensorflow/lite/tools/flatbuffer_utils.py (hunk 12 above)
Comment (kun-lu20) [reply tensorflow/lite/tools/flatbuffer_utils.py]: Updated. Thanks!

File: tensorflow/lite/tools/flatbuffer_utils.py (hunk 13 above)
Comment (kun-lu20) [reply tensorflow/lite/tools/flatbuffer_utils.py]: Added description for the remaining arguments. Thanks!

File: tensorflow/lite/tools/flatbuffer_utils.py (hunk 14 above)
Comment (kun-lu20) [reply tensorflow/lite/tools/flatbuffer_utils.py]: Updated. Thanks!

File: tensorflow/lite/tools/flatbuffer_utils.py (hunk 15 above)
Comment (kun-lu20) [reply tensorflow/lite/tools/flatbuffer_utils.py]: Updated. Thanks!

File: tensorflow/lite/tools/flatbuffer_utils.py [hunk 16]
```diff
@@ -274,6 +281,78 @@ def xxd_output_to_object(input_cc_file):
...
+
+  Args:
+    tflite_model: TFLite flatbuffer in a byte array of from_endiness format.
```
[the remark this answers is not in the document]
Comment (kun-lu20) [reply tensorflow/lite/tools/flatbuffer_utils.py]: Description added. Thanks!

**Comment by kun-lu20** [comment]
Hi @terryheo ,

Thanks for pointing out the above issues. I've addressed them and please take a look when you have a chance. Thanks again!

**Review by terryheo — APPROVED** [review]
LGTM
